frame/state-trie-migration/src/lib.rs

```diff
@@ -0,0 +1,1525 @@
+				task.migrate_until_exhaustion(SignedMigrationMaxLimits::get());
+
+				frame_support::assert_ok!(StateTrieMigration::continue_migrate(
+					Origin::signed(1),
+					SignedMigrationMaxLimits::get(),
+					task.dyn_size
+				));
+
+				// no funds should remain reserved.
+				assert_eq!(Balances::reserved_balance(&1), 0);
+
+				// and the task should be updated
+				assert!(matches!(
+					StateTrieMigration::migration_process(),
+					MigrationTask { size: x, .. } if x > 0,
+				));
+			}
+		});
+	}
+
+	#[test]
+	fn custom_migrate_works() {
+		new_test_ext(StateVersion::V0, true).execute_with(|| {
+			frame_support::assert_ok!(StateTrieMigration::migrate_custom_top(
+				Origin::signed(1),
+				vec![b"key1".to_vec(), b"key2".to_vec(), b"key3".to_vec()],
+				10 + 20 + 30,
+			));
+
+			// no funds should remain reserved.
+			assert_eq!(Balances::reserved_balance(&1), 0);
+		});
+	}
+}
+
+#[cfg(all(test, feature = "remote-tests"))]
+mod remote_tests {
+	use super::{mock::*, *};
+	use codec::Encode;
+	use mock::run_to_block_and_drain_pool;
+	use remote_externalities::{Mode, OfflineConfig, OnlineConfig};
+	use sp_io::InMemoryProvingBackend;
+	use sp_runtime::traits::{Bounded, HashFor};
+	use std::sync::Arc;
+
+	// we only use the hash type from this, so using the mock should be fine.
+	type Block = sp_runtime::testing::Block<Extrinsic>;
+
+	#[tokio::test]
+	async fn on_initialize_migration() {
+		sp_tracing::try_init_simple();
+		let run_with_limits = |limits| async move {
+			let mut ext = remote_externalities::Builder::<Block>::new()
+				.mode(Mode::OfflineOrElseOnline(
+					OfflineConfig {
+						state_snapshot: "/home/kianenigma/remote-builds/state".to_owned().into(),
+					},
+					OnlineConfig {
+						transport: std::env!("WS_API").to_owned().into(),
+						scrape_children: true,
+						state_snapshot: Some(
+							"/home/kianenigma/remote-builds/state".to_owned().into(),
+						),
+						..Default::default()
+					},
+				))
+				.state_version(sp_core::StateVersion::V0)
+				.build()
+				.await
+				.unwrap();
+
+			let mut now = ext.execute_with(|| {
+				AutoLimits::<Test>::put(Some(limits));
+				// requires the block number type in our tests to be same as with mainnet, u32.
+				frame_system::Pallet::<Test>::block_number()
+			});
+
+			let mut duration = 0;
+			// set the version to 1, as if the upgrade happened.
+			ext.state_version = sp_core::StateVersion::V1;
+
+			loop {
+				let trie_backend = ext.backend.clone();
+				let last_state_root = trie_backend.root().clone();
+				let proving_backend = InMemoryProvingBackend::new(&trie_backend);
+				let (finished, proof) = ext.execute_and_get_proof(&proving_backend, || {
+					run_to_block(now + 1);
+					if StateTrieMigration::migration_process().finished() {
+						return true
+					}
+					duration += 1;
+					now += 1;
+					false
+				});
+
+				let (top_left, child_left) =
+					ext.as_backend().essence().check_migration_state().unwrap();
+				log::info!(
```

Comment:
Would be interesting to have in the log, the size of the proof before the call to into_compact_proof.
Another interesting size would be the size of the compacted proof after being compressed by zstd.

Reply:
great idea: 


```
Nov 18 19:04:20.076  INFO runtime::state-trie-migration: [7759753] 🤖 migrated 32768 top keys, 0 child keys, and a total of 2416621 bytes.
Nov 18 19:04:24.795  INFO runtime::state-trie-migration: proceeded to #7759753, original proof: 5379156, compact proof size: 3980046, compact zstd compressed: 1795656 // top_left: 819977, child_left: 63874
Nov 18 19:04:27.761  INFO runtime::state-trie-migration: [7759754] 🤖 migrated 32768 top keys, 0 child keys, and a total of 2621440 bytes.
Nov 18 19:04:32.096  INFO runtime::state-trie-migration: proceeded to #7759754, original proof: 5806331, compact proof size: 4406130, compact zstd compressed: 1844988 // top_left: 787209, child_left: 63874
Nov 18 19:04:35.054  INFO runtime::state-trie-migration: [7759755] 🤖 migrated 32768 top keys, 0 child keys, and a total of 2621440 bytes.
Nov 18 19:04:39.402  INFO runtime::state-trie-migration: proceeded to #7759755, original proof: 5811741, compact proof size: 4407584, compact zstd compressed: 1844440 // top_left: 754441, child_left: 63874
Nov 18 19:04:42.343  INFO runtime::state-trie-migration: [7759756] 🤖 migrated 32768 top keys, 0 child keys, and a total of 2621440 bytes.
Nov 18 19:04:46.719  INFO runtime::state-trie-migration: proceeded to #7759756, original proof: 5811785, compact proof size: 4406041, compact zstd compressed: 1844469 // top_left: 721673, child_left: 63874
Nov 18 19:04:49.701  INFO runtime::state-trie-migration: [7759757] 🤖 migrated 32768 top keys, 0 child keys, and a total of 2621440 bytes.
```